ArticleTemplates/assets/js/bootstraps/campaign.js

```diff
@@ -29,26 +51,44 @@ function initCampaign(campaign) {
 
     var form = campaign.querySelector('form');
     form.addEventListener('submit', function (e) {
+        displayWaiting(form);
         e.preventDefault();
         var data = Array.from(form.elements).reduce(function (o, e) {
             if (e.type === 'checkbox') {
                 if (e.checked) {
                     o[e.name] = o[e.name] ? o[e.name] + '\n' + e.value : e.value;
                 }
+            } else if (e.type === 'file') {
+                o[e.name] = readFile(e.files[0], campaign, form);
             } else if (e.value) {
                 o[e.name] = e.value;
             }
             
             return o;    
         }, {});
-        disableButton(form);
-        submit(data, campaign, form);
+
+        let promises = [];
+        let keys = [];
+        for (var key in data) {
+            if (data.hasOwnProperty(key) && data[key] instanceof Promise) {
+                keys.push(key)
+                promises.push(data[key]);
+            }
+        }
+
+        Promise.all(promises).then(results => {
+            results.map((result, index) => {
+                data[keys[index]] = result;
+            })
+
+            disableButton(form);
```

Comment:
shouldn't the button be disabled before we try to read the files asynchronously?